packages/library-sync/src/commonMain/kotlin/io/realm/kotlin/mongodb/Functions.kt

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2022 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.kotlin.mongodb
+
+import io.realm.kotlin.mongodb.exceptions.FunctionExecutionException
+import io.realm.kotlin.mongodb.internal.BsonEncoder
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.serializer
+import kotlinx.serialization.serializerOrNull
+import org.mongodb.kbson.BsonArray
+import org.mongodb.kbson.BsonDocument
+import org.mongodb.kbson.BsonValue
+import kotlin.reflect.typeOf
+
+/**
+ * A Functions manager to call remote Atlas Functions for the associated Atlas App services Application.
+ *
+ * Since the serialization engine [does not support third-party libraries yet](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md), there are some
+ * limitations in what types can be used as arguments and return types:
+ *
+ * - Primitives, Bson, lists, and maps are valid argument types.
+ * - Results can only be deserialized to primitives or Bson types.
+ *
+ * @see [User.functions]
+ */
+public interface Functions {
+    /**
+     * The [App] that this function manager is associated with.
+     */
+    public val app: App
+
+    /**
+     * The [User] that this function manager is authenticated with.
+     */
+    public val user: User
+
+    /**
+     * Invokes an Atlas function.
+     *
+     * Since the serialization engine [does not support third-party libraries yet](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/formats.md), there are some
+     * limitations in what types can be used as arguments and return types:
+     *
+     * - Primitives, Bson, lists, and maps are valid argument types.
+     * - Results can only be deserialized to primitives or Bson types.
+     *
+     * The Bson implementations for arrays or maps are [BsonArray] and [BsonDocument], and they can be
+     * used as valid return types.
+     *
+     * @param name Name of the function to call.
+     * @param args Arguments to the function.
+     * @param deserializationStrategy Deserialization strategy for decoding the results.
+     * @param T The type for the functions response.
+     * @return Result of the function.
+     *
+     * @throws FunctionExecutionException if the request failed in some way.
```

Comment:
```suggestion
     * @throws FunctionExecutionException if the function failed in some way.
```

Reply:
We should probably also add 

```
     * @throws io.realm.kotlin.mongodb.exceptions.ServiceException for other failures that can happen when
     * communicating with App Services. See [AppException] for details.
```

Similar to what we do in ApiKeyAuth and EmailPasswordAuth